File: sunnad-ios/sunnad-ios/Domain/Models/HabitCompletion.swift
import Foundation

enum CompletionEntrySource: String, Codable, Hashable, Sendable {
    case normal
    case lateCheckIn = "late_check_in"
}

struct HabitCompletion: Codable, Hashable, Sendable {
    var habitID: UUID
    var dayDate: Date
    var value: Int
    var completedAt: Date?
    var updatedAt: Date
    var entrySource: CompletionEntrySource

    init(
        habitID: UUID,
        dayDate: Date,
        value: Int,
        completedAt: Date? = nil,
        updatedAt: Date = Date(),
        entrySource: CompletionEntrySource = .normal
    ) {
        self.habitID = habitID
        self.dayDate = dayDate
        self.value = value
        self.completedAt = completedAt
        self.updatedAt = updatedAt
        self.entrySource = entrySource
    }

    func isValid(for habit: Habit) -> Bool {
        switch habit.type {
        case .binary:
            return value == 0 || value == 1
        case .dhikr:
            return value >= 0
        }
    }

    func isCompleted(for habit: Habit) -> Bool {
        switch habit.type {
        case .binary:
            return value == 1
        case .dhikr:
            return value >= habit.normalizedTargetCount
        }
    }

    func clamped(for habit: Habit) -> HabitCompletion {
        var copy = self

        switch habit.type {
        case .binary:
            copy.value = value <= 0 ? 0 : 1
        case .dhikr:
            copy.value = max(value, 0)
        }

        return copy
    }
}

File: sunnad-android/app/src/main/java/com/arystan/almasuly/sunnadandroid/core/model/HabitCompletion.kt
package com.arystan.almasuly.sunnadandroid.core.model

import java.time.Instant
import java.time.LocalDate
import java.util.UUID

data class HabitCompletion(
    val habitId: UUID,
    val dayDate: LocalDate,
    val value: Int,
    val completedAt: Instant? = null,
    val updatedAt: Instant = Instant.now()
) {
    fun isValid(forHabit: Habit): Boolean {
        return when (forHabit.type) {
            HabitType.BINARY -> value == 0 || value == 1
            HabitType.DHIKR -> value in 0..forHabit.normalizedTargetCount
        }
    }

    fun isCompleted(forHabit: Habit): Boolean {
        return when (forHabit.type) {
            HabitType.BINARY -> value == 1
            HabitType.DHIKR -> value >= forHabit.normalizedTargetCount
        }
    }

    fun clamped(forHabit: Habit): HabitCompletion {
        val normalizedValue = when (forHabit.type) {
            HabitType.BINARY -> if (value <= 0) 0 else 1
            HabitType.DHIKR -> value.coerceIn(0, forHabit.normalizedTargetCount)
        }
        return copy(value = normalizedValue)
    }
}
